lib/IRGen/GenBuiltin.cpp

```diff
@@ -643,6 +643,23 @@ if (Builtin.ID == BuiltinValueKind::id) { \
     return;
   }
 
+  if (Builtin.ID == BuiltinValueKind::StackAlloc) {
+    auto size = args.claimNext();
+    auto align = args.claimNext();    
+
+    if (IGF.isAsync() || IGF.isCoroutine()) {
+      // These types of functions don't actually stack-allocate, and support for
+      // them is not needed at this time.
+      IGF.IGM.error(SourceLoc(), "Builtin.stackAlloc() is not implemented for async functions.");
+      out.add(llvm::Constant::getNullValue(IGF.IGM.Int8Ty));
+      return;
+    }
+
+    auto alloc = IGF.emitDynamicAlloca(IGF.IGM.Int8Ty, size, align, Alignment(16), "builtin-stackAlloc");
```

Comment:
So if the align isn't a compile time constant, you'll default to 16 bytes?

I think we could maybe develop a better API here. What if we had two overloads: one that takes an alignment argument and one that doesn't. If the alignment argument is provided and isn't a compile time constant you could create an error. This also means you wouldn't have to create another overload of `emitDynamicAlloca`.

Reply:
It is intentional that the parameters to the calling API might be determined at runtime. The selection of 16-byte alignment is intentional: it's the default alignment for heap-allocated memory in Swift as well. Of course, we might want to fall back to something else or actually honour the alignment request where possible—I don't know LLVM well enough to know how to go about building that though.

Reply:
(This is a WIP PR I opened so I could run tests. It needs to be reviewed both in swift-evolution and by the core team before it can hope to be merged.)

Reply:
FWIW I'm working on new codegen here that should act a bit smarter.